### PR TITLE
Added stop committee runner queue consumption trigger

### PR DESCRIPTION
**Pull request by AKorpusenko**

### CODE REVIEW

File: protocol/v2/ssv/runner/committee.go [hunk 1]
```diff
@@ -43,6 +43,8 @@ type CommitteeRunner struct {
 	started       time.Time
 	consensusDone time.Time
 	postStarted   time.Time
+
+	stopCh chan struct{}
```
Comment (y0sher) [inline protocol/v2/ssv/runner/committee.go:47]: why is this here if it has no meaning for the committee runner itself?

File: protocol/v2/ssv/validator/committee_queue.go [hunk 2]
```diff
@@ -72,7 +72,12 @@ func (v *Committee) HandleMessage(logger *zap.Logger, msg *queue.DecodedSSVMessa
 
 // ConsumeQueue consumes messages from the queue.Queue of the controller
 // it checks for current state
-func (v *Committee) ConsumeQueue(logger *zap.Logger, slot phase0.Slot, handler MessageHandler) error {
+func (v *Committee) ConsumeQueue(
+	logger *zap.Logger,
+	slot phase0.Slot,
+	handler MessageHandler,
+	stopCh <-chan struct{},
```
Comment (y0sher) [inline protocol/v2/ssv/validator/committee_queue.go:79]: should've probably just used a context dedicated to this slot queue.

**Review by y0sher — CHANGES_REQUESTED** [review]
create a context when you create the queue and pass it in.

File: protocol/v2/ssv/validator/committee_queue.go [hunk 3]
```diff
@@ -72,11 +72,32 @@ func (v *Committee) HandleMessage(logger *zap.Logger, msg *queue.DecodedSSVMessa
...
+	defer func() {
+		v.mtx.Lock()
+		defer v.mtx.Unlock()
+
+		var ok bool
+		q, ok = v.Queues[slot]
+		if !ok {
+			logger.Warn("committee queue not found for slot while stopping the consumer", fields.Slot(slot))
+			return
+		}
```
[the remark this answers is not in the document]
Comment (AKorpusenko) [reply protocol/v2/ssv/validator/committee_queue.go]: Added. Thanks

File: protocol/v2/ssv/validator/committee_queue.go [hunk 4]
```diff
@@ -72,11 +72,33 @@ func (v *Committee) HandleMessage(logger *zap.Logger, msg *queue.DecodedSSVMessa
...
 	var q queueContainer
+
+	// in case of any error try to call the ctx.cancel to prevent the ctx leak
+	defer func() {
+		var ok bool
+
+		v.mtx.Lock()
+		q, ok = v.Queues[slot]
+		v.mtx.Unlock()
+
```
Comment (y0sher) [inline protocol/v2/ssv/validator/committee_queue.go:90]: `StartDuty` is already locking, so this is potential "deadlock" or just double locking, although this happens in a goroutine and the lock is released in start duty. Still I think its better to just pass the Queue `q` from `StartDuty` as we're already using the mutex to get it there.

by passing it to this function you don't need to use the mutex here at all.

File: operator/validator/controller.go [hunk 5]
```diff
@@ -864,6 +864,9 @@ func (c *controller) onShareStop(pubKey spectypes.ValidatorPK) {
 	if ok {
 		vc.RemoveShare(v.Share.Share.ValidatorIndex)
 		if len(vc.Shares) == 0 {
+			for _, slotQueue := range vc.Queues {
+				slotQueue.StopQueueF()
+			}
```
Comment (y0sher) [inline operator/validator/controller.go:869]: you can't do this here, accessing Queues requires mutex acquiring. best is to have a function inside the committee struct that handles this.

**Review by y0sher — CHANGES_REQUESTED** [review]
concurrency issues..

File: operator/validator/controller.go [hunk 6]
```diff
@@ -864,6 +864,7 @@ func (c *controller) onShareStop(pubKey spectypes.ValidatorPK) {
 	if ok {
 		vc.RemoveShare(v.Share.Share.ValidatorIndex)
 		if len(vc.Shares) == 0 {
+			vc.StopQueues()
```
Comment (y0sher) [inline operator/validator/controller.go:867]: `Stop` or `Close` is more appropriate

File: protocol/v2/ssv/validator/events.go [hunk 7]
```diff
@@ -42,13 +43,22 @@ func (c *Committee) handleEventMessage(logger *zap.Logger, msg *queue.DecodedSSV
 		if err != nil {
 			return err
 		}
+
 		c.mtx.Lock()
-		dutyRunner := c.Runners[slot] // TODO: err check , runner exist?
+		q := c.Queues[slot]
+		dutyRunner, exists := c.Runners[slot]
 		c.mtx.Unlock()
+		if !exists {
```
Comment (y0sher) [inline protocol/v2/ssv/validator/events.go:51]:
```suggestion
		q, exists := c.Queues[slot]
		dutyRunner, exists := c.Runners[slot]
		c.mtx.Unlock()
		if !exists || !qexists {
```

queue should be present if there's a runner but lets be on the safe side anyway.

**Review by y0sher — CHANGES_REQUESTED** [review]
small change

**Review by y0sher — APPROVED** [review]

File: protocol/v2/ssv/validator/events.go [hunk 8]
```diff
@@ -42,13 +43,23 @@ func (c *Committee) handleEventMessage(logger *zap.Logger, msg *queue.SSVMessage
...
+		if !rExists || !qExists {
+			logger.Error("no committee runner or queue found for slot", fields.Slot(slot), fields.MessageID(msg.MsgID))
+			return nil
+		}
+
 		if err := dutyRunner.GetBaseRunner().QBFTController.OnTimeout(logger, *eventMsg); err != nil {
 			return fmt.Errorf("timeout event: %w", err)
 		}
+
+		q.StopQueueF()
```
Comment (moshe-blox) [inline protocol/v2/ssv/validator/events.go:61]: why are we stopping the queue on timeout message?

**Comment by AKorpusenko** [comment]
Replaced with PR https://github.com/ssvlabs/ssv/pull/1556 based on the `stage` branch

**Pull request closed without merging**